Source/cmake/OptionsWPE.cmake

```diff
@@ -452,5 +453,26 @@ else ()
     set(WPEWebProcessExtension_Uninstalled_PKGCONFIG_FILE ${CMAKE_BINARY_DIR}/wpe-web-extension-${WPE_API_VERSION}-uninstalled.pc)
 endif ()
 
+if (USE_SIZE_OPTIMIZATIONS)
+    if (COMPILER_IS_GCC_OR_CLANG AND UNIX AND NOT APPLE)
+        if (CMAKE_COMPILER_IS_GNUCC)
+            # finline-limit is supported on GCC only.
+            set(CMAKE_COMPILER_SIZE_OPT_FLAGS " -finline-limit=90")
+            # fsection-anchors is supported on ARM/ARM64 and not on X86 or X86-64.
+            # It may be supported on other platforms, but the GCC documentation doesn't specify it, other
+            # than saying that the feature is platform dependant.
+            if (WTF_CPU_ARM64 OR WTF_CPU_ARM)
+                set(CMAKE_COMPILER_SIZE_OPT_FLAGS "${CMAKE_COMPILER_SIZE_OPT_FLAGS} -fsection-anchors")
+            endif ()
+        endif ()
+        set(CMAKE_C_FLAGS_RELEASE "${CMAKE_C_FLAGS_RELEASE}${CMAKE_COMPILER_SIZE_OPT_FLAGS} -ffunction-sections -fdata-sections")
+        set(CMAKE_CXX_FLAGS_RELEASE "${CMAKE_CXX_FLAGS_RELEASE}${CMAKE_COMPILER_SIZE_OPT_FLAGS} -ffunction-sections -fdata-sections")
+        set(CMAKE_SHARED_LINKER_FLAGS_RELEASE "${CMAKE_SHARED_LINKER_FLAGS_RELEASE} -Wl,--gc-sections")
+        set(CMAKE_C_FLAGS_DEBUG "${CMAKE_C_FLAGS_DEBUG}${CMAKE_COMPILER_SIZE_OPT_FLAGS} -ffunction-sections -fdata-sections")
+        set(CMAKE_CXX_FLAGS_DEBUG "${CMAKE_CXX_FLAGS_DEBUG}${CMAKE_COMPILER_SIZE_OPT_FLAGS} -ffunction-sections -fdata-sections")
+        set(CMAKE_SHARED_LINKER_FLAGS_DEBUG "${CMAKE_SHARED_LINKER_FLAGS_DEBUG} -Wl,--gc-sections")
+    endif ()
```

Comment:
You could replace all this with a single `WEBKIT_APPEND_GLOBAL_COMPILER_FLAGS` invocation, except for the linker flags. For the linker flags I would add a `LD_SUPPORTS_GC_SECTIONS` variable in `OptionsCommon.cmake` (we already have some `LD_SUPPORTS_<option>` there), and skip using the variables with the `_RELEASE` and `_DEBUG` suffixes to use the flags in all kinds of builds. Lastly, I would also set the executable and loadable shared module flags.

Like this:

```suggestion
    WEBKIT_APPEND_GLOBAL_COMPILER_FLAGS(-finline-limit=90
                                        -fsection-anchors)
    if (LD_SUPPORTS_GC_SECTIONS)
        WEBKIT_APPEND_GLOBAL_COMPILER_FLAGS(-ffunction-sections
                                            -fdata-sections)
        set(CMAKE_EXE_LINK_FLAGS "${CMAKE_EXE_LINKER_FLAGS} -Wl,--gc-sections")
        set(CMAKE_SHARED_LINKER_FLAGS "${CMAKE_SHARED_LINKER_FLAGS} -Wl,--gc-sections")
        set(CMAKE_MODULE_LINKER_FLAGS "${CMAKE_SHARED_LINKER_FLAGS} -Wl,--gc-sections")
    endif ()
```

Like @mcatanzaro mentioned, this should probably go in `WebKitCompilerFlags.cmake` or in `OptionsCommon.cmake`.

I think we are reaching a point of linker flag tweaking that would justify having helper CMake functions to check whether the linker supports a given flag, and another to pass a list of flags and automatically add the ones supported by the linker. But that could be a follow-up patch.

Reply:
FWIW, I think it does _not_ make sense to apply the flags only to `MinSizeRel` builds. One may want an optimized build and at the same time do the `--gc-sections` part&mdash;because it always provides a gain.

OTOH, options `-finline-limit=` and `-fsection-anchors` may produce less performant code, so those we would want only for `MinSizeRel` builds so these two makes sense to add to `CMAKE_{C,CXX}_FLAGS_MINSIZEREL`. Recently I did a refactor in our CMake compiler-flag functions which allows doing the same:

```cmake
WEBKIT_VAR_ADD_COMPILER_FLAGS(C APPEND CMAKE_C_FLAGS_MINSIZEREL
    -finline-limit=90 -fsection-anchors)
WEBKIT_VAR_ADD_COMPILER_FLAGS(CXX APPEND CMAKE_CXX_FLAGS_MINSIZEREL
    -finline-limit=90 -fsection-anchors)
```

Only those options that are supported by the compiler will be appended to the variables.